File: native/sen3/FileFixes/t0100_dat.cpp
﻿#include <filesystem>
#include <string_view>
#include <vector>

#include "../file_fixes.h"

#include "../../p3a/pack.h"
#include "../../p3a/structs.h"
#include "../../sha1.h"

namespace SenLib::Sen3::FileFixes::t0100_dat {
std::string_view GetDescription() {
    return "Textbox overflow in Recette Bakery.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/t0100.dat",
            56609,
            SenPatcher::SHA1FromHexString("02b743e33e402065f67af2fa318064f41314d353"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        // chapter 1 system message for going to the einhel keep
        // swap linebreak forwards so line doesn't overflow textbox
        std::swap(bin[0x8a0a], bin[0x8a19]);

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::t0100_dat

File: SenLib/Sen3/FileFixes/t0100_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class t0100_dat : FileMod {
		public string GetDescription() {
			return "Textbox overflow in Recette Bakery.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x02b743e33e402065ul, 0xf67af2fa318064f4ul, 0x1314d353u));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();

			// chapter 1 system message for going to the einhel keep
			// swap linebreak forwards so line doesn't overflow textbox
			bin.SwapBytes(0x8a0a, 0x8a19);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/t0100.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x02b743e33e402065ul, 0xf67af2fa318064f4ul, 0x1314d353u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/t0100.dat", file) };
		}
	}
}
